archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,872 @@
+        lt_add_iff_pos_left, canonically_ordered_comm_semiring.mul_pos, succ_pos', and_self], },
+    have h_geom := has_sum_geometric_of_lt_1 h_nonneg hlt,
+    have h_geom' := has_sum.mul_left ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) h_geom,
+    norm_num at h_geom',
+    norm_num at h_pow_succ,
+    have h_geom'' :
+      has_sum (λ (b : ℕ), (1 / ((2 * ((n : ℝ) + 1) + 1) ^ 2) ^ b.succ))
+      (1 / (2 * ((n : ℝ) + 1) + 1) ^ 2 * (1 - 1 / (2 * ((n : ℝ) + 1) + 1) ^ 2)⁻¹) :=
+    begin
+      refine has_sum.has_sum_of_sum_eq _ h_geom',
+      intros,
+      use u,
+      intros v' h,
+      use v',
+      split,
+      { exact h },
+      { refine sum_congr rfl _, intros k hk, exact h_pow_succ k },
+    end,
+    norm_num,
+    exact h_geom'',
+  end,
+  have hab :
+    ∀ (k : ℕ), (1 / (2 * (k.succ : ℝ) + 1)) * ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ ≤
+    ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ :=
+  begin
+    intro k,
+    have h_zero_le : 0 ≤ ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ, by
+    { simp [cast_succ, one_div, inv_pow, inv_nonneg], norm_cast, exact zero_le', },
+    have h_left : 1 / (2 * (k.succ : ℝ) + 1) ≤ 1, by
+    { simp only [cast_succ, one_div],
+      refine inv_le_one _,
+      norm_cast,
+      exact (le_add_iff_nonneg_left 1).mpr zero_le', },
+    exact mul_le_of_le_one_left h_zero_le h_left,
+  end,
+  exact has_sum_le hab (log_stirling_seq_diff_has_sum n) g,
+end
+
+/--
+We have the bound  `log_stirling_seq n - log_stirling_seq (n+1)` ≤ 1/(4n(n+1))
+-/
+lemma log_stirling_seq_sub_log_stirling_seq_succ : ∀ (n : ℕ),
+log_stirling_seq n.succ - log_stirling_seq n.succ.succ ≤ 1 / (4 * n.succ * n.succ.succ) :=
+begin
+  intro n,
+  have h₁ : 0 < 4 * (n.succ : ℝ) * (n.succ.succ : ℝ), by
+  { norm_cast, simp only [canonically_ordered_comm_semiring.mul_pos, succ_pos', and_self], },
+  have h₂ : 0 < 1 - (1 / (2 * (n.succ : ℝ) + 1)) ^ 2 :=
+  begin
+    refine sub_pos.mpr _,
+    refine (sq_lt_one_iff _).mpr _,
+    { rw one_div,
+      refine inv_nonneg.mpr _,
+      norm_cast,
+      exact zero_le (2 * succ n + 1) },
+    { refine (div_lt_one _).mpr _,
+      all_goals {norm_cast},
+      linarith,
+      refine lt_add_of_pos_left 1 ((1 : ℕ).succ_mul_pos (succ_pos n)), },
+  end,
+  refine le_trans (log_stirling_seq_diff_le_geo_sum n) ((le_div_iff' h₁).mpr _),
+  rw mul_div (4 * (n.succ : ℝ) * (n.succ.succ : ℝ))
+    ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) (1 - (1 / (2 * (n.succ : ℝ) + 1)) ^ 2),
+  refine (div_le_one h₂).mpr _,
+  norm_num,
+  rw mul_div,
+  have h₃: 0 < (2 * ((n : ℝ) + 1) + 1) ^ 2 :=
+  begin
+    rw sq,
+    norm_cast,
+    simp only [canonically_ordered_comm_semiring.mul_pos,
+    succ_pos', and_self],
+  end,
+  refine (div_le_iff' h₃).mpr _,
+  rw mul_sub,
+  norm_num,
+  refine le_sub.mp _,
+  rw mul_one_div,
+  refine (div_le_iff' h₃).mpr ((le_mul_iff_one_le_right h₃).mpr
+    (le_sub_iff_add_le.mpr _)),
+  norm_cast,
+  rw sq,
+  linarith,
+end
```

Comment:
```suggestion
lemma log_stirling_seq_sub_log_stirling_seq_succ (n : ℕ) :
  log_stirling_seq n.succ - log_stirling_seq n.succ.succ ≤ 1 / (4 * n.succ * n.succ.succ) :=
begin
  have h₁ : 0 < 4 * ((n:ℝ) + 1) * (((n:ℝ) + 1) + 1) := by nlinarith [@cast_nonneg ℝ _ n],
  have h₃ : 0 < (2 * ((n:ℝ) + 1) + 1) ^ 2 := by linarith,
  have h₂ : 0 < 1 - (1 / (2 * ((n:ℝ) + 1) + 1)) ^ 2,
  { rw ← mul_lt_mul_right h₃,
    have H : 0 < (2 * ((n:ℝ) + 1) + 1) ^ 2 - 1 := by linarith,
    convert H using 1; field_simp [h₃.ne'] },
  refine le_trans (log_stirling_seq_diff_le_geo_sum n) _,
  push_cast at *,
  rw div_le_div_iff h₂ h₁,
  rw ← mul_le_mul_right h₃,
  have H : 4 * ((n:ℝ) + 1) * (n + 1 + 1) ≤ (2 * (↑n + 1) + 1) ^ 2 - 1 := by linarith,
  convert H using 1; field_simp [h₃.ne'],
end
```
Another golf.  The lesson here is that mathlib has very good tools for inequalities (`(n)linarith`) and for denominator-cancelling (`field_simp`), but can't really mix them.  It's fastest to reorganize so that the steps of the two types are kept separate.